gRPC connection fail when `--join` with TLS enabled

Please answer these questions before submitting your issue. Thanks!

1. What did you do?

When trying to add TLS support to tidb-operator ([#750](https://github.com/pingcap/tidb-operator/pull/750)), we found that `--join` is not working correctly with TLS enabled.

Assume there're `pd-0`, `pd-1` and `pd-2`, starting sequently. `pd-0` starts as normal with `--initial-cluster=pd-0=https://pd-0:2380` and TLS enabled, and `pd-1` starts with `--join=https://pd-0:2380`, then `pd-1` is not able to connect to `pd-0` and exits.

The errors from `pd-1` are:
```
[2019/08/14 05:50:43.611 -04:00] [ERROR] [join.go:180] ["failed to open directory"] [error="open /home/tidb/deploy/pd-server/member: no such file or directory"]
2019/08/14 05:50:43.612 grpclog.go:45: [info] parsed scheme: "endpoint"
2019/08/14 05:50:43.612 grpclog.go:45: [info] ccResolverWrapper: sending new addresses to cc: [{https://172.16.5.146:2380 0  <nil>}]
2019/08/14 05:50:43.613 grpclog.go:60: [warning] grpc: addrConn.createTransport failed to connect to {https://172.16.5.146:2380 0  <nil>}. Err :connection error: desc = "transport: Error while dialing dial tcp 172.16.5.146:2380: connect: connection refused". Reconnecting...
2019/08/14 05:50:44.613 grpclog.go:60: [warning] grpc: addrConn.createTransport failed to connect to {https://172.16.5.146:2380 0  <nil>}. Err :connection error: desc = "transport: Error while dialing dial tcp 172.16.5.146:2380: connect: connection refused". Reconnecting...
2019/08/14 05:50:46.138 grpclog.go:60: [warning] grpc: addrConn.createTransport failed to connect to {https://172.16.5.146:2380 0  <nil>}. Err :connection error: desc = "transport: Error while dialing dial tcp 172.16.5.146:2380: connect: connection refused". Reconnecting...
2019/08/14 05:50:48.301 grpclog.go:60: [warning] grpc: addrConn.createTransport failed to connect to {https://172.16.5.146:2380 0  <nil>}. Err :connection error: desc = "transport: Error while dialing dial tcp 172.16.5.146:2380: connect: connection refused". Reconnecting...
2019/08/14 05:50:52.802 grpclog.go:60: [warning] grpc: addrConn.createTransport failed to connect to {https://172.16.5.146:2380 0  <nil>}. Err :connection error: desc = "transport: Error while dialing dial tcp 172.16.5.146:2380: connect: connection refused". Reconnecting...
[2019/08/14 05:50:53.613 -04:00] [FATAL] [main.go:85] ["join meet error"] [error="context deadline exceeded"] [stack="github.com/pingcap/log.Fatal\n\t/home/jenkins/workspace/release_tidb_3.0/go/pkg/mod/github.com/pingcap/log@v0.0.0-20190715063458-479153f07ebd/global.go:59\nmain.main\n\t/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/pd/cmd/pd-server/main.go:85\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:200"]
```

And there errors on `pd-0` shows:
```
[2019/08/14 05:57:27.602 -04:00] [WARN] [grpclog.go:60] ["grpc: Server.Serve failed to complete security handshake from \"172.16.5.147:59118\": tls: first record does not look like a TLS handshake"]
```
So this might because `pd-1` uses a plain connection to try to connect to `pd-0`, instead of a TLS encrypted connection.

We initially found this issue on a Kubernetes deployment, but it's also reproducible with binary deployment, the procedure are as above:
 - deploy a cluster using tidb-ansible with `enable_tls = True` in `inventory.ini`, but do not start it
 - login to `pd-0`, change `pd-0:/path/to/deploy/script/run_pd.sh` and set `--initial-cluster=pd-0=https://pd-0:2380` (remove other nodes), and start it with `sudo systemctl start pd-2379.service`
 - login to `pd-1`, change `pd-1:/path/to/deploy/script/run_pd.sh` and delete `--initial-cluster` line, add a `--join=https://pd-0:2380` line, then start it with `sudo systemctl start pd-2379.service`
 - watch `pd-0` and `pd-1`'s log

2. What did you expect to see?

The process work normally

3. What did you see instead?

`pd-1` produce errors and exits

4. What version of PD are you using (`pd-server -V`)?
```
Release Version: v3.0.1
Git Commit Hash: 811ce0b9a1335d1b2a049fd97ef9e186f1c9efc1
Git Branch: HEAD
UTC Build Time:  2019-07-16 01:02:23
```
This issue also exists with `v3.0.2` docker image.

--- a/client/client.go
+++ b/client/client.go
@@ -15,10 +15,6 @@ package pd
 
 import (
 	"context"
-	"crypto/tls"
-	"crypto/x509"
-	"io/ioutil"
-	"net/url"
 	"strings"
 	"sync"
 	"time"
@@ -27,10 +23,10 @@ import (
 	"github.com/pingcap/kvproto/pkg/metapb"
 	"github.com/pingcap/kvproto/pkg/pdpb"
 	"github.com/pingcap/log"
+	"github.com/pingcap/pd/pkg/grpcutil"
 	"github.com/pkg/errors"
 	"go.uber.org/zap"
 	"google.golang.org/grpc"
-	"google.golang.org/grpc/credentials"
 )
 
 // Client is a PD (Placement Driver) client.
@@ -272,43 +268,7 @@ func (c *client) getOrCreateGRPCConn(addr string) (*grpc.ClientConn, error) {
 		return conn, nil
 	}
 
-	opt := grpc.WithInsecure()
-	if len(c.security.CAPath) != 0 {
-
-		certificates := []tls.Certificate{}
-		if len(c.security.CertPath) != 0 && len(c.security.KeyPath) != 0 {
-			// Load the client certificates from disk
-			certificate, err := tls.LoadX509KeyPair(c.security.CertPath, c.security.KeyPath)
-			if err != nil {
-				return nil, errors.Errorf("could not load client key pair: %s", err)
-			}
-			certificates = append(certificates, certificate)
-		}
-
-		// Create a certificate pool from the certificate authority
-		certPool := x509.NewCertPool()
-		ca, err := ioutil.ReadFile(c.security.CAPath)
-		if err != nil {
-			return nil, errors.Errorf("could not read ca certificate: %s", err)
-		}
-
-		// Append the certificates from the CA
-		if !certPool.AppendCertsFromPEM(ca) {
-			return nil, errors.New("failed to append ca certs")
-		}
-
-		creds := credentials.NewTLS(&tls.Config{
-			Certificates: certificates,
-			RootCAs:      certPool,
-		})
-
-		opt = grpc.WithTransportCredentials(creds)
-	}
-	u, err := url.Parse(addr)
-	if err != nil {
-		return nil, errors.WithStack(err)
-	}
-	cc, err := grpc.Dial(u.Host, opt)
+	cc, err := grpcutil.GetClientConn(addr, c.security.CAPath, c.security.CertPath, c.security.KeyPath)
 	if err != nil {
 		return nil, errors.WithStack(err)
 	}

--- a/pkg/grpcutil/grpcutil.go
+++ b/pkg/grpcutil/grpcutil.go
@@ -1,0 +1,69 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package grpcutil
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"io/ioutil"
+	"net/url"
+
+	"github.com/pkg/errors"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+)
+
+// GetClientConn returns a gRPC client connection.
+func GetClientConn(addr string, caPath string, certPath string, keyPath string) (*grpc.ClientConn, error) {
+	opt := grpc.WithInsecure()
+	if len(caPath) != 0 {
+		certificates := []tls.Certificate{}
+		if len(certPath) != 0 && len(keyPath) != 0 {
+			// Load the client certificates from disk
+			certificate, err := tls.LoadX509KeyPair(certPath, keyPath)
+			if err != nil {
+				return nil, errors.Errorf("could not load client key pair: %s", err)
+			}
+			certificates = append(certificates, certificate)
+		}
+
+		// Create a certificate pool from the certificate authority
+		certPool := x509.NewCertPool()
+		ca, err := ioutil.ReadFile(caPath)
+		if err != nil {
+			return nil, errors.Errorf("could not read ca certificate: %s", err)
+		}
+
+		// Append the certificates from the CA
+		if !certPool.AppendCertsFromPEM(ca) {
+			return nil, errors.New("failed to append ca certs")
+		}
+
+		creds := credentials.NewTLS(&tls.Config{
+			Certificates: certificates,
+			RootCAs:      certPool,
+		})
+
+		opt = grpc.WithTransportCredentials(creds)
+	}
+	u, err := url.Parse(addr)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	cc, err := grpc.Dial(u.Host, opt)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	return cc, nil
+}

--- a/server/region_syncer/client.go
+++ b/server/region_syncer/client.go
@@ -15,20 +15,15 @@ package syncer
 
 import (
 	"context"
-	"crypto/tls"
-	"crypto/x509"
-	"io/ioutil"
-	"net/url"
 	"time"
 
 	"github.com/pingcap/kvproto/pkg/pdpb"
 	"github.com/pingcap/log"
+	"github.com/pingcap/pd/pkg/grpcutil"
 	"github.com/pingcap/pd/server/core"
 	"github.com/pkg/errors"
 	"go.uber.org/zap"
-	"google.golang.org/grpc"
 	"google.golang.org/grpc/codes"
-	"google.golang.org/grpc/credentials"
 	"google.golang.org/grpc/status"
 )
 
@@ -56,46 +51,9 @@ func (s *RegionSyncer) reset() {
 func (s *RegionSyncer) establish(addr string) (ClientStream, error) {
 	s.reset()
 
-	u, err := url.Parse(addr)
+	cc, err := grpcutil.GetClientConn(addr, s.securityConfig.CAPath, s.securityConfig.CertPath, s.securityConfig.KeyPath)
 	if err != nil {
-		return nil, err
-	}
-
-	opt := grpc.WithInsecure()
-	if len(s.securityConfig.CAPath) != 0 {
-		certificates := []tls.Certificate{}
-		if len(s.securityConfig.CertPath) != 0 && len(s.securityConfig.KeyPath) != 0 {
-			// Load the client certificates from disk
-			certificate, e := tls.LoadX509KeyPair(s.securityConfig.CertPath, s.securityConfig.KeyPath)
-			if e != nil {
-				return nil, errors.Errorf("could not load client key pair: %s", e)
-			}
-			certificates = append(certificates, certificate)
-		}
-
-		// Create a certificate pool from the certificate authority
-		certPool := x509.NewCertPool()
-		ca, e := ioutil.ReadFile(s.securityConfig.CAPath)
-		if e != nil {
-			return nil, errors.Errorf("could not read ca certificate: %s", e)
-		}
-
-		// Append the certificates from the CA
-		if !certPool.AppendCertsFromPEM(ca) {
-			return nil, errors.New("failed to append ca certs")
-		}
-
-		creds := credentials.NewTLS(&tls.Config{
-			Certificates: certificates,
-			RootCAs:      certPool,
-		})
-
-		opt = grpc.WithTransportCredentials(creds)
-	}
-
-	cc, err := grpc.Dial(u.Host, opt, grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(msgSize)))
-	if err != nil {
-		return nil, err
+		return nil, errors.WithStack(err)
 	}
 
 	ctx, cancel := context.WithCancel(s.server.Context())

--- a/server/region_syncer/client.go
+++ b/server/region_syncer/client.go
@@ -15,15 +15,20 @@ package syncer
 
 import (
 	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"io/ioutil"
 	"net/url"
 	"time"
 
 	"github.com/pingcap/kvproto/pkg/pdpb"
 	"github.com/pingcap/log"
 	"github.com/pingcap/pd/server/core"
+	"github.com/pkg/errors"
 	"go.uber.org/zap"
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials"
 	"google.golang.org/grpc/status"
 )
 
@@ -56,7 +61,39 @@ func (s *RegionSyncer) establish(addr string) (ClientStream, error) {
 		return nil, err
 	}
 
-	cc, err := grpc.Dial(u.Host, grpc.WithInsecure(), grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(msgSize)))
+	opt := grpc.WithInsecure()
+	if len(s.securityConfig.CAPath) != 0 {
+		certificates := []tls.Certificate{}
+		if len(s.securityConfig.CertPath) != 0 && len(s.securityConfig.KeyPath) != 0 {
+			// Load the client certificates from disk
+			certificate, e := tls.LoadX509KeyPair(s.securityConfig.CertPath, s.securityConfig.KeyPath)
+			if e != nil {
+				return nil, errors.Errorf("could not load client key pair: %s", e)
+			}
+			certificates = append(certificates, certificate)
+		}
+
+		// Create a certificate pool from the certificate authority
+		certPool := x509.NewCertPool()
+		ca, e := ioutil.ReadFile(s.securityConfig.CAPath)
+		if e != nil {
+			return nil, errors.Errorf("could not read ca certificate: %s", e)
+		}
+
+		// Append the certificates from the CA
+		if !certPool.AppendCertsFromPEM(ca) {
+			return nil, errors.New("failed to append ca certs")
+		}
+
+		creds := credentials.NewTLS(&tls.Config{
+			Certificates: certificates,
+			RootCAs:      certPool,
+		})
+
+		opt = grpc.WithTransportCredentials(creds)
+	}
+
+	cc, err := grpc.Dial(u.Host, opt, grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(msgSize)))
 	if err != nil {
 		return nil, err
 	}

--- a/server/region_syncer/server.go
+++ b/server/region_syncer/server.go
@@ -23,6 +23,7 @@ import (
 	"github.com/pingcap/kvproto/pkg/metapb"
 	"github.com/pingcap/kvproto/pkg/pdpb"
 	"github.com/pingcap/log"
+	"github.com/pingcap/pd/server/config"
 	"github.com/pingcap/pd/server/core"
 	"github.com/pkg/errors"
 	"go.uber.org/zap"
@@ -59,19 +60,21 @@ type Server interface {
 	GetStorage() *core.Storage
 	Name() string
 	GetMetaRegions() []*metapb.Region
+	GetSecurityConfig() *config.SecurityConfig
 }
 
 // RegionSyncer is used to sync the region information without raft.
 type RegionSyncer struct {
 	sync.RWMutex
-	streams map[string]ServerStream
-	ctx     context.Context
-	cancel  context.CancelFunc
-	server  Server
-	closed  chan struct{}
-	wg      sync.WaitGroup
-	history *historyBuffer
-	limit   *ratelimit.Bucket
+	streams        map[string]ServerStream
+	ctx            context.Context
+	cancel         context.CancelFunc
+	server         Server
+	closed         chan struct{}
+	wg             sync.WaitGroup
+	history        *historyBuffer
+	limit          *ratelimit.Bucket
+	securityConfig *config.SecurityConfig
 }
 
 // NewRegionSyncer returns a region syncer.
@@ -81,11 +84,12 @@ type RegionSyncer struct {
 // no longer etcd but go-leveldb.
 func NewRegionSyncer(s Server) *RegionSyncer {
 	return &RegionSyncer{
-		streams: make(map[string]ServerStream),
-		server:  s,
-		closed:  make(chan struct{}),
-		history: newHistoryBuffer(defaultHistoryBufferSize, s.GetStorage().GetRegionStorage()),
-		limit:   ratelimit.NewBucketWithRate(defaultBucketRate, defaultBucketCapacity),
+		streams:        make(map[string]ServerStream),
+		server:         s,
+		closed:         make(chan struct{}),
+		history:        newHistoryBuffer(defaultHistoryBufferSize, s.GetStorage().GetRegionStorage()),
+		limit:          ratelimit.NewBucketWithRate(defaultBucketRate, defaultBucketCapacity),
+		securityConfig: s.GetSecurityConfig(),
 	}
 }
 

